### PR TITLE
NIAD-1054: Handle out of order and illogical messages

### PR DESCRIPTION
**Pull request by Broom94**

### CODE REVIEW

File: service/src/main/java/uk/nhs/adaptors/gp2gp/mhs/InboundMessageHandler.java [hunk 1]
```diff
@@ -57,6 +59,9 @@ public boolean handle(Message message) {
                 );
             }
             return true;
+        } catch (EhrExtractMessageOutOfOrderException | EhrExtractNonExistingException | UnsupportedInteractionException e) {
+            LOGGER.error("An error occurred while handing MHS inbound message {}", messageID, e);
+            throw e;
```
[the remark this answers is not in the document]
Comment (Broom94) [reply service/src/main/java/uk/nhs/adaptors/gp2gp/mhs/InboundMessageHandler.java:64]: Removed log keeping rethrow, I was always told this is best practice to rethrow error to root

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/EhrExtractStatusService.java [hunk 2]
```diff
@@ -385,4 +415,44 @@ private Update createUpdateWithUpdatedAt() {
 
         return update;
     }
+
+    private boolean checkForAckOutOfOrderAndDuplicate(String conversationId) {
+        var ehrExtractStatusOptional = ehrExtractStatusRepository.findByConversationId(conversationId);
+        if (ehrExtractStatusOptional.isPresent()) {
+            EhrExtractStatus ehrExtractStatus = ehrExtractStatusOptional.get();
```
Comment (msolecka) [inline service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/EhrExtractStatusService.java:422]: Optional, fetching status could be refactored with orElseThrow

``` 
EhrExtractStatus ehrExtractStatus = ehrExtractStatusRepository.findByConversationId(conversationId)
            .orElseThrow(() -> new EhrExtractNonExistingException("ACK", conversationId));
            
```

the same in checkForContinueOutOfOrderAndDuplicate.

File: service/build.gradle [hunk 3]
```diff
@@ -60,6 +60,7 @@ dependencies {
 	implementation 'javax.xml.soap:javax.xml.soap-api:1.4.0'
 	implementation 'com.github.spullara.mustache.java:compiler:0.9.5'
 	implementation 'org.apache.tika:tika-core:1.26'
+	implementation 'org.awaitility:awaitility:4.0.3'
```
Comment (msolecka) [inline service/build.gradle:63]: Do we need this dependency? it is only used once in test and it has been already added as test dependency.

**Review by msolecka — DISMISSED** [review]

File: service/src/main/java/uk/nhs/adaptors/gp2gp/mhs/InboundMessageHandler.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (lgonczar) [reply service/src/main/java/uk/nhs/adaptors/gp2gp/mhs/InboundMessageHandler.java:64]: A part of this method's responsibility is to catch any kind of error that occurs during the processing of the message, handle the situation (e.g. mark the process as failed) and return a result indicating whether we've processed the message (i.e. if we're done with it). Logging + returning `false` would be consistent with its current behaviour.

[the remark this answers is not in the document]
Comment (mattd-kainos) [reply service/src/main/java/uk/nhs/adaptors/gp2gp/mhs/InboundMessageHandler.java:63]: @lgonczar is right here, this is the "root" e.g. top level place where all errors are caught and handled.

[the remark this answers is not in the document]
Comment (Broom94) [reply service/src/main/java/uk/nhs/adaptors/gp2gp/mhs/InboundMessageHandler.java:63]: I think it is the InboundMessageConsumer no? As this class calls this inboundmessagehandler

[the remark this answers is not in the document]
Comment (mattd-kainos) [reply service/src/main/java/uk/nhs/adaptors/gp2gp/mhs/InboundMessageHandler.java:63]: I see, I guess the unusual part of this code is that we "catch and release" a specific subset of exception but catch-all and handle the general case.

Should handleMessageProcessingError and all exception handling be moved down into the InboundMessageConsumer then?

**Review by msolecka — DISMISSED** [review]

File: service/src/main/java/uk/nhs/adaptors/gp2gp/mhs/InboundMessageHandler.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (mattd-kainos) [reply service/src/main/java/uk/nhs/adaptors/gp2gp/mhs/InboundMessageHandler.java:63]: I haven't worked in this part of the code in a while so perhaps my feedback isn't the best. I'll leave it to both of your to collaborate and find the correct solution here.

[the remark this answers is not in the document]
Comment (Broom94) [reply service/src/main/java/uk/nhs/adaptors/gp2gp/mhs/InboundMessageHandler.java]: returned false

File: service/src/main/java/uk/nhs/adaptors/gp2gp/mhs/InboundMessageHandler.java [hunk 4]
```diff
@@ -78,7 +79,7 @@ private InboundMessage unmarshallMessage(Message message) {
         }
     }
 
-    private void handleInboundMessage(ParsedInboundMessage inboundMessage) {
+    public void handleInboundMessage(ParsedInboundMessage inboundMessage) {
```
Comment (lgonczar) [inline service/src/main/java/uk/nhs/adaptors/gp2gp/mhs/InboundMessageHandler.java:82]: Why is this method public?

File: service/src/intTest/java/uk/nhs/adaptors/gp2gp/ehr/IllogicalMessageComponentTest.java [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+    private static final String TO_ODS_CODE_PATH = SUBJECT_PATH + "/EhrRequest/destination/AgentOrgSDS/agentOrganizationSDS/id/@extension";
+    private static final String MESSAGE_ID_PATH = MESSAGE_HEADER_PATH + "/MessageData/MessageId";
+    private static final String EHR_EXTRACT_REQUEST = "RCMR_IN010000UK05";
+    private static final String CONTINUE_REQUEST = "COPC_IN000001UK01";
+    private static final String ACKNOWLEDGMENT_REQUEST = "MCCI_IN010000UK13";
+    private static final String ACTION_PATH = "/Envelope/Header/MessageHeader/Action";
+    private static final String CONVERSATION_ID_PATH = "/Envelope/Header/MessageHeader/ConversationId";
+    private static final String NON_EXISTING_CONVERSATION_ID = "d3746650-096e-414b-92a4-146ceaf74f0e";
+    private static final String ACK_OK_CODE = "AA";
+    private static final XPathService SERVICE = new XPathService();
```
Comment (lgonczar) [inline service/src/intTest/java/uk/nhs/adaptors/gp2gp/ehr/IllogicalMessageComponentTest.java:73]: Worth giving this service a more specific name.

**Review by lgonczar — APPROVED** [review]

File: service/src/intTest/java/uk/nhs/adaptors/gp2gp/ehr/IllogicalMessageComponentTest.java [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+import uk.nhs.adaptors.gp2gp.mhs.InboundMessageHandler;
+import uk.nhs.adaptors.gp2gp.mhs.exception.UnsupportedInteractionException;
+import uk.nhs.adaptors.gp2gp.testcontainers.ActiveMQExtension;
+import uk.nhs.adaptors.gp2gp.testcontainers.MongoDBExtension;
+
+@RunWith(SpringRunner.class)
+@ExtendWith({SpringExtension.class, MongoDBExtension.class, ActiveMQExtension.class, MockitoExtension.class})
+@SpringBootTest
+@DirtiesContext
+public class IllogicalMessageComponentTest {
```
Comment (lgonczar) [inline service/src/intTest/java/uk/nhs/adaptors/gp2gp/ehr/IllogicalMessageComponentTest.java:53]: Nice tests 👍

**Review by lgonczar — APPROVED** [review]

**Review by lgonczar — APPROVED** [review]

**Review by msolecka — APPROVED** [review]

**Pull request merged**